File: scripts/deploy.py
#!/usr/bin/env python3
"""Deploy sakana-2api to server via paramiko (SFTP)."""
import paramiko, os, sys, tarfile, io

import json as _json
try:
    with open(os.path.join(os.path.dirname(__file__), '.ssh_secret.json')) as _f:
        _sec = _json.load(_f)
except FileNotFoundError:
    _sec = {}
HOST = os.environ.get('SAKANA_SSH_HOST', _sec.get('HOST', '38.76.190.150'))
PORT = int(os.environ.get('SAKANA_SSH_PORT', _sec.get('PORT', 22)))
USER = os.environ.get('SAKANA_SSH_USER', _sec.get('USER', 'root'))
PASS = os.environ.get('SAKANA_SSH_PASS', _sec.get('PASS', ''))
LOCAL = r'D:\workspaces\sakana-2api'
REMOTE = '/root/sakana-2api'

ssh = paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
ssh.connect(HOST, PORT, USER, PASS, look_for_keys=False, allow_agent=False)
print(f'Connected to {HOST}')

# Create remote dir
ssh.exec_command(f'mkdir -p {REMOTE}')

# Upload via tar-pipe through SFTP
sftp = ssh.open_sftp()

# Build tar in memory, excluding temp files
buf = io.BytesIO()
tar = tarfile.open(fileobj=buf, mode='w:gz')
exclude = {'.git', 'node_modules', 'session.json', 'tokens.json', 'server.log', 'raw_search.txt', 'raw_search_full.txt', 'red.png', 'red64.png', 'chunk-1ceo.js', 'chunk-messages.js', 'mailmsg.json', 'mailtoken.json', 'tempmail.json', 'success_sample.json', 'capture_', 'scan_', 'probe_', 'replicate_', 'test_', 'dump_', 'check_', 'ui_upload', 'brute_stream', 'tamper_test', 'net_trace', 'debug_stream', 'harvest_session', 'find_bundle', 'firebase_login', 'node_replay_sample', 'open_magic_link', 'click_login', 'compare_auth', 'complete_login', 'do_login', 'capture_', 'chunk-'}
for root, dirs, files in os.walk(LOCAL):
    for dn in list(dirs):
        if dn in exclude or dn.startswith('.'):
            dirs.remove(dn)
    for fn in files:
        if any(fn.startswith(p) for p in exclude):
            continue
        if fn.endswith(('.exe', '.png', '.txt')) and not fn in ('README.md', 'requirements.txt'):
            continue
        local = os.path.join(root, fn)
        remote = os.path.relpath(local, LOCAL)
        try:
            tar.add(local, arcname=remote)
        except Exception as e:
            print(f'  skip {remote}: {e}')
tar.close()
buf.seek(0)

# Write tar to remote via SFTP
with sftp.open(f'{REMOTE}/project.tar.gz', 'wb') as f:
    f.write(buf.read())
print('Uploaded project.tar.gz')

# Extract remotely
stdin, stdout, stderr = ssh.exec_command(f'cd {REMOTE} && tar xzf project.tar.gz && rm project.tar.gz && ls -la')
print(stdout.read().decode())
err = stderr.read().decode()
if err:
    print('stderr:', err[:500])

# Test
stdin, stdout, stderr = ssh.exec_command(f'cd {REMOTE} && node server.js &')
import time
time.sleep(2)
stdin, stdout, stderr = ssh.exec_command(f'curl -s -m 10 http://127.0.0.1:8787/health')
print('Health:', stdout.read().decode())

stdin, stdout, stderr = ssh.exec_command(f'cat {REMOTE}/server.log')
print('Log:', stdout.read().decode()[-500:])

print('Done!')
ssh.close()

File: scripts/deploy.mjs
// Deploy script: SSH into server, install Node/Chrome, clone repo, start service
import { spawn } from 'child_process';
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs';

const fs = require('fs');
const path = require('path');
// Credentials from scripts/.ssh_secret.json (gitignored) or env vars.
const _secPath = path.join(__dirname, '.ssh_secret.json');
const _sec = fs.existsSync(_secPath) ? JSON.parse(fs.readFileSync(_secPath, 'utf8')) : {};
const HOST = process.env.SAKANA_SSH_HOST || _sec.HOST || '38.76.190.150';
const PORT = Number(process.env.SAKANA_SSH_PORT || _sec.PORT || 22);
const USER = process.env.SAKANA_SSH_USER || _sec.USER || 'root';
const PASS = process.env.SAKANA_SSH_PASS || _sec.PASS || '';
const REMOTE_DIR = '/root/sakana-2api';

function sshExec(cmd, opts = {}) {
  return new Promise((resolve, reject) => {
    const args = ['-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null',
      '-p', String(PORT), `${USER}@${HOST}`, cmd];
    const proc = spawn('ssh', args, { stdio: ['pipe', 'pipe', 'pipe'], ...opts });
    let out = '', err = '';
    proc.stdout.on('data', d => out += d.toString());
    proc.stderr.on('data', d => {
      const s = d.toString();
      err += s;
      // Send password when prompted
      if (s.toLowerCase().includes('password:')) {
        proc.stdin.write(PASS + '\n');
      }
    });
    proc.on('close', code => {
      if (code === 0) resolve({ out, err });
      else reject(new Error(`exit ${code}: ${err.slice(0, 500)}`));
    });
    proc.on('error', reject);
  });
}

async function deploy() {
  console.log('=== 1. Check server environment ===');
  try {
    const r = await sshExec('uname -a && echo "---NODE---" && node --version 2>/dev/null || echo "NODE_NOT_FOUND" && echo "---CHROME---" && (google-chrome --version 2>/dev/null || chromium --version 2>/dev/null || echo "CHROME_NOT_FOUND") && echo "---MEM---" && free -h | grep Mem');
    console.log(r.out);
  } catch (e) {
    // First connection might fail on password prompt timing
    console.log('Initial connect attempt:', e.message.slice(0, 100));
  }

  console.log('\n=== 2. Install Node.js if missing ===');
  try {
    const r = await sshExec('which node && node --version');
    if (r.out.includes('NODE_NOT_FOUND') || !r.out.includes('v')) {
      console.log('Installing Node.js...');
      await sshExec('curl -fsSL https://deb.nodesource.com/setup_22.x | bash - && apt-get install -y nodejs');
    } else {
      console.log('Node.js already installed:', r.out.trim());
    }
  } catch (e) { console.log('Node check:', e.message.slice(0, 100)); }

  console.log('\n=== 3. Create project directory & upload files ===');
  // We'll use a tar-pipe approach: tar the local project, pipe over SSH, extract
  await sshExec(`mkdir -p ${REMOTE_DIR}`);

  // Tar-pipe the project (excluding .git, node_modules, temp files)
  const tarProc = spawn('tar', ['czf', '-',
    '--exclude=.git', '--exclude=node_modules', '--exclude=session.json',
    '--exclude=tokens.json', '--exclude=server.log',
    '--exclude=*.png',
    '-C', '/d/workspaces/sakana-2api', '.']);
  
  const sshProc = spawn('ssh', ['-o', 'StrictHostKeyChecking=no', '-p', String(PORT),
    `${USER}@${HOST}`, `tar xzf - -C ${REMOTE_DIR}`],
    { stdio: ['pipe', 'pipe', 'pipe'] });
  
  let sshErr = '';
  let passwordSent = false;
  sshProc.stderr.on('data', d => {
    const s = d.toString();
    sshErr += s;
    if (s.toLowerCase().includes('password:') && !passwordSent) {
      passwordSent = true;
      sshProc.stdin.write(PASS + '\n');
    }
  });
  tarProc.stdout.pipe(sshProc.stdin);
  tarProc.stdout.on('end', () => { sshProc.stdin.end(); });

  await new Promise((resolve, reject) => {
    sshProc.on('close', code => {
      if (code === 0) resolve();
      else reject(new Error(`upload exit ${code}: ${sshErr.slice(0, 300)}`));
    });
    sshProc.on('error', reject);
  });
  console.log('Files uploaded successfully');

  console.log('\n=== 4. Install dependencies & start ===');
  console.log('(If Node/npm not found, install first)');
  await sshExec(`cd ${REMOTE_DIR} && npm install --production 2>&1 || true`);
  const start = await sshExec(`cd ${REMOTE_DIR} && nohup node server.js > server.log 2>&1 & sleep 2 && cat server.log`);
  console.log('Start result:', start.out);

  console.log('\n=== 5. Verify ===');
  const verify = await sshExec(`curl -s -m 10 http://127.0.0.1:8787/health`);
  console.log('Health check:', verify.out);

  console.log('\n✅ Deployment complete!');
  console.log(`API endpoint: http://${HOST}:8787/v1/chat/completions`);
}

deploy().catch(e => console.error('Deploy failed:', e.message));